### ISSUE TITLE
`prod(())` shouldn't return `1`

### ISSUE THREAD

**Issue by jariji** [JuliaLang/julia#55675]
```jl
julia> prod(())
1

julia> prod([])
ERROR: MethodError: no method matching one(::Type{Any})
```

It's guessing that the output type is `Int`, which is bad. I think this should throw an error.

**Comment by mbauman** [comment]
Theoretically I agree, but pragmatically this method is often used for size tuples. This change would be surprisingly breaking for precisely that reason.

**Comment by jariji** [comment]
Ok then maybe a breaking change for 2.0.

**Comment by barucden** [comment]
@mbauman 
> this method is often used for size tuples

As in `prod(size(A))`? Maybe Julia should provide a function for getting the number of elements in a collection then?

This issue is, IMHO, quite similar to #55379 in that Julia returns an (arbitrary) value for an empty collection.

**Comment by mbauman** [comment]
I dunno, someone could try it and see just how breaking it is, but unlike #55379, this isn't giving objectively _wrong_ answers.  That tends to happen when we guess at `init` — this is guessing at the empty case.

**Comment by DNF2** [comment]
> As in `prod(size(A))`? Maybe Julia should provide a function for getting the number of elements in a collection then?

That's what `length(A)` does.

**Comment by barucden** [comment]
Indeed! I never realized `length` does this. I guess I always assumed it returned the number of elements along the first dimension (influenced by python's `len` perhaps).

What's the motivation for `prod(size(A))` then? [People use it.](https://juliahub.com/ui/Search?q=prod%5C(size%5C(.%2B%5C)%5C)&type=code&rx=true) 

Is it true that any `prod(size(A))` can be replaced by `length(A)`?

I believe the pattern `prod(size(A)[c:d]` cannot be replaced by `length`, but `size(A)[c:d]` should never return `()`, unless `c>d`, which seems like a very marginal case.

**Comment by nsajko** [comment]
For any empty collection `c`, `isone(prod(c))` must hold, right? Given that, returning `1` doesn't seem *that* bad. It's not ideal for element types that don't subtype `Number`, though.

**Comment by barucden** [comment]
Oh, I see now that I completely missed what OP was reporting! My apologies, I'll hide my previous comments. Let me make it clear for any other confused contributors.

The issue is not returning the multiplicative identity for an empty collection. That is the standard definition in math. And programming languages (R, Common Lisp, numpy) follow that definition, including Julia:
```julia
julia> prod(Int[])
1
```
The documentation for `prod` says that:
> The value returned for empty itr can be specified by init. It must be the multiplicative identity (i.e. one) as it is unspecified whether init is used for non-empty collections.



The issue is *guessing what the multiplicative identity is for an empty tuple*. We do not guess it for `Any[]`, but we guess it for `Tuple(Any[])`.
```julia
julia> A = []
Any[]

julia> prod(A)
ERROR: MethodError: no method matching one(::Type{Any})

julia> prod(Tuple(A))
1
```
